File: src-tauri/src/brand.rs
//! Single source of truth for every brand-bearing identifier that crosses a
//! persistence or wire boundary — Rust side. Mirrors `src/shared/brand.ts`.
//!
//! Two constant groups, and the distinction between them is load-bearing:
//!
//! - [`LEGACY`] — the values already on users' disks / in their agent memories.
//!   Permanent. Only ever *read*; nothing may write them again (FORBID-04).
//!   This file and `src/shared/brand.ts` are the only two non-fixture files
//!   permitted to contain a legacy brand string.
//! - [`canonical`] — the values the app writes *today*. Renaming a contract is
//!   therefore a one-line edit here, which is what makes "no repo-wide sed" a
//!   structural property rather than a slogan.
//!
//! [`canonical`] reads through an override seam so a harness test can inject
//! the post-rename value while production still emits the pre-rename one. That
//! is how a Wave-1 red can be a *real* red rather than a self-certifying
//! assertion (`assert_eq!(CONST, "same literal")` can never fail).
//!
//! The seam is `pub` rather than `#[cfg(test)]` because the harness lives in
//! `src-tauri/tests/`, which links this crate as an external dependency and so
//! never sees `cfg(test)` items. Production code must never call it; the
//! residual-scan gate greps for stray call sites.

use std::cell::Cell;


/// Every canonical brand identifier the app writes.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct BrandCanonical {
    /// `ConversationRecordV2.createdBy` discriminant.
    pub created_by: &'static str,
    /// Markdown fence language carrying an ACP plan payload.
    pub plan_fence: &'static str,
    /// Per-user-repository workspace directory name.
    pub workspace_dir: &'static str,
    /// Short display name; also the `~/Documents/<name>` path component.
    pub display_name: &'static str,
    /// Full display name / product name.
    pub display_name_full: &'static str,
    /// Desktop binary and cargo package name.
    pub package_name: &'static str,
    /// Standalone headless server binary name.
    pub server_binary: &'static str,
    /// Production bundle identifier.
    pub bundle_id: &'static str,
    /// Development bundle identifier.
    pub bundle_id_dev: &'static str,
    /// Base file name (no extension) of the desktop log file.
    pub log_file_name: &'static str,
    /// `log`/`tracing` target prefix and `RUST_LOG` directive stem.
    pub log_target: &'static str,
    /// Keychain service holding desktop general credentials.
    pub keychain_service: &'static str,
    /// Keychain service holding SSH passwords and key passphrases.
    pub keychain_ssh_service: &'static str,
    /// App-managed `~/.ssh/` host-key store, kept separate from the user's own
    /// `known_hosts` so libssh2's writer cannot drop `@cert-authority` /
    /// `@revoked` markers it does not understand.
    ///
    /// DISCOVERED DURING EXECUTE — absent from all 14 roots in
    /// `migration-plan.json` and from the analyze inventory. Renaming this file
    /// without migrating it does not fail loudly: the host-key store simply
    /// looks empty, every known host becomes "unknown", and `accept-new`
    /// silently re-trusts whatever answers — which is precisely the state a
    /// MITM needs. It must be migrated, not just renamed.
    pub ssh_known_hosts_file: &'static str,
    /// Keychain service holding the iOS pairing secret.
    pub keychain_pairing_service: &'static str,
    /// MCP server name exposed to agents.
    pub mcp_server_name: &'static str,
    /// Name of the managed scheduled-tasks agent skill.
    pub skill_name: &'static str,
    /// HTML marker identifying a skill file this app wrote.
    pub skill_marker: &'static str,
    /// On-disk key of the managed-skill manifest's ownership flag, in
    /// `<workspace_dir>/managed-skills.json`.
    ///
    /// `ManagedSkillManifestV1` carries `#[serde(rename_all = "camelCase",
    /// deny_unknown_fields)]`, so the Rust field identifier *is* this JSON key —
    /// there is no literal anywhere to grep for, and renaming the identifier is
    /// a one-token edit that compiles and silently rewrites an external
    /// contract. `deny_unknown_fields` is what makes that data loss rather than
    /// drift: a manifest already on a user's disk stops deserializing outright
    /// the instant the key moves.
    ///
    /// T-A21 moved it and bumped the manifest's `schema_version` 1 -> 2. The
    /// pre-rename key survives as a permanent `#[serde(alias)]` — the one place
    /// outside this file and `src/shared/brand.ts` allowed to spell a legacy
    /// brand value, because serde attributes take literals only and cannot read
    /// this constant. The agreement between that attribute and this constant is
    /// guarded by `tests/legacy_brand_skill_manifest.rs` rather than by a
    /// whitelist entry nothing enforces.
    pub skill_manifest_key: &'static str,
    /// frp `[[proxies]]` registration name.
    pub frp_proxy_name: &'static str,
    /// Standalone-server state root directory name (unix, lowercase).
    pub state_dir: &'static str,
    /// Standalone-server state root directory name (Windows, capitalized).
    pub state_dir_windows: &'static str,
    /// Negotiated binary WebSocket subprotocol.
    pub ws_subprotocol: &'static str,
    /// Prefix of every environment variable the app reads.
    pub env_prefix: &'static str,
    /// Prefix of every global/element id injected into third-party pages.
    pub dom_global_prefix: &'static str,
    /// Deep-link URL scheme (without `://`).
    pub deep_link_scheme: &'static str,
    /// `tauri-plugin-store` namespace prefix — `<prefix><namespace>::<key>` in
    /// the WebView's `localStorage`. Mirrors `storagePrefix` in
    /// `src/shared/brand.ts`.
    pub storage_prefix: &'static str,
    /// Prefix of every bare app-owned `localStorage` key the renderer writes
    /// outside the store plugin. Mirrors `storageKeyPrefix` in
    /// `src/shared/brand.ts`.
    ///
    /// Both prefixes exist on the Rust side for one reason: on macOS the
    /// WebView data store is partitioned by bundle identifier and cannot be
    /// moved, so the app has to read its own keys out under the old identifier
    /// and replay them under the new one — see
    /// `src/webview_storage_handoff.rs`.
    pub storage_key_prefix: &'static str,
}

/// Values already written to user disks. Permanent, read-only, never re-emitted.
///
/// Migration and compatibility-read paths are the *only* legitimate consumers.
pub const LEGACY: BrandCanonical = BrandCanonical {
    created_by: "termul",
    plan_fence: "termul-plan",
    workspace_dir: ".termul",
    display_name: "Termul",
    display_name_full: "Termul Manager",
    package_name: "termul-manager",
    server_binary: "termul-server",
    bundle_id: "com.termul-manager.app",
    bundle_id_dev: "com.termul-manager.app.dev",
    log_file_name: "termul",
    log_target: "termul",
    keychain_service: "com.termul.manager",
    keychain_ssh_service: "termul-ssh",
    ssh_known_hosts_file: "known_hosts_termul",
    keychain_pairing_service: "com.termul.remote.pairing",
    mcp_server_name: "termul",
    skill_name: "termul-scheduled-tasks",
    skill_marker: "<!-- managed-by-termul:termul-scheduled-tasks -->",
    skill_manifest_key: "managedByTermul",
    frp_proxy_name: "termul",
    state_dir: "termul",
    state_dir_windows: "Termul",
    ws_subprotocol: "termul-terminal-v2.binary",
    env_prefix: "TERMUL_",
    dom_global_prefix: "__termul",
    deep_link_scheme: "termul",
    storage_prefix: "termul-store:",
    storage_key_prefix: "termul:",
};

/// Values the app writes today.
///
/// Wave 5 flips these one contract at a time. Until a contract's flip task
/// lands, its entry here still equals the corresponding [`LEGACY`] value — that
/// gap is precisely what makes the Wave-1 harness tests go red.
pub const DEFAULT_CANONICAL: BrandCanonical = BrandCanonical {
    created_by: "se-manager",
    plan_fence: "termul-plan",
    workspace_dir: ".termul",
    display_name: "Se",
    display_name_full: "Termul Manager",
    package_name: "termul-manager",
    server_binary: "termul-server",
    bundle_id: "com.termul-manager.app",
    bundle_id_dev: "com.termul-manager.app.dev",
    log_file_name: "termul",
    log_target: "termul",
    keychain_service: "com.se-manager.app",
    keychain_ssh_service: "com.se-manager.ssh",
    ssh_known_hosts_file: "known_hosts_se-manager",
    keychain_pairing_service: "com.termul.remote.pairing",
    mcp_server_name: "se-manager",
    skill_name: "se-manager-scheduled-tasks",
    skill_marker: "<!-- managed-by-se-manager:se-manager-scheduled-tasks -->",
    skill_manifest_key: "managedBySeManager",
    frp_proxy_name: "se-manager",
    state_dir: "termul",
    state_dir_windows: "Termul",
    ws_subprotocol: "se-terminal-v2.binary",
    env_prefix: "TERMUL_",
    dom_global_prefix: "__se",
    deep_link_scheme: "termul",
    storage_prefix: "se-store:",
    storage_key_prefix: "se:",
};

thread_local! {
    /// Per-thread override. Cargo runs tests in parallel threads inside one
    /// process, so a *process*-global seam would leak one test's injected value
    /// into every sibling test. Thread-local is therefore the correct default.
    static THREAD_OVERRIDE: Cell<Option<BrandCanonical>> = const { Cell::new(None) };
}

/// The canonical brand values in force right now.
///
/// Always call this rather than caching the result in a `static` — a cached
/// value freezes before a test can override it.
pub fn canonical() -> BrandCanonical {
    THREAD_OVERRIDE
        .with(Cell::get)
        .unwrap_or(DEFAULT_CANONICAL)
}

/// Test seam: force canonical values on **this thread** until the guard drops.
///
/// Production never calls this. Harness tests use it to inject the *post*-rename
/// value while production still emits the pre-rename one, so the resulting red
/// reflects a real missing capability instead of a stale literal.
#[doc(hidden)]
#[must_use = "the override is reverted when the guard is dropped"]
pub fn override_canonical(next: BrandCanonical) -> BrandOverrideGuard {
    let previous = THREAD_OVERRIDE.with(|slot| slot.replace(Some(next)));
    BrandOverrideGuard { previous }
}

/// Reverts an [`override_canonical`] call when dropped.
#[doc(hidden)]
pub struct BrandOverrideGuard {
    previous: Option<BrandCanonical>,
}

impl Drop for BrandOverrideGuard {
    fn drop(&mut self) {
        THREAD_OVERRIDE.with(|slot| slot.set(self.previous));
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn canonical_defaults_to_shipped_values() {
        assert_eq!(canonical(), DEFAULT_CANONICAL);
    }

    #[test]
    fn override_replaces_canonical_and_reverts_on_drop() {
        let before = canonical();
        {
            let _guard = override_canonical(BrandCanonical {
                created_by: "se-manager",
                ..DEFAULT_CANONICAL
            });
            assert_eq!(canonical().created_by, "se-manager");
            // Untouched fields still come from the shipped values.
            assert_eq!(canonical().plan_fence, DEFAULT_CANONICAL.plan_fence);
        }
        assert_eq!(canonical(), before);
    }

    #[test]
    fn legacy_values_are_independent_of_the_override_seam() {
        let _guard = override_canonical(BrandCanonical {
            created_by: "se-manager",
            workspace_dir: ".se-manager",
            ..DEFAULT_CANONICAL
        });
        assert_eq!(LEGACY.created_by, "termul");
        assert_eq!(LEGACY.workspace_dir, ".termul");
    }

    /// The whole harness runs under `cargo test`'s parallel thread pool. If one
    /// test's injected value were visible to a sibling, every red in Wave 1
    /// would be non-deterministic and the ledger would prove nothing.
    #[test]
    fn override_does_not_leak_into_other_threads() {
        let _guard = override_canonical(BrandCanonical {
            created_by: "se-manager",
            ..DEFAULT_CANONICAL
        });
        assert_eq!(canonical().created_by, "se-manager");

        let observed = std::thread::spawn(|| canonical().created_by).join().unwrap();
        assert_eq!(observed, DEFAULT_CANONICAL.created_by);
    }
}

File: src/shared/brand.ts
/**
 * Single source of truth for every brand-bearing identifier that crosses a
 * persistence or wire boundary.
 *
 * Two constant groups, and the distinction between them is load-bearing:
 *
 * - `LEGACY_*` — the values that are already on users' disks / in their agent
 *   memories. They are **permanent**. They are only ever *read*; nothing in the
 *   app may write them again (FORBID-04). This module is the only legitimate
 *   home for a legacy brand literal outside the frozen fixture roots.
 * - `CANONICAL_*` — the values the app writes *today*. Renaming a contract is
 *   therefore a one-line edit here rather than a 400-file replacement, which is
 *   what makes "no repo-wide sed" a structural property instead of a slogan.
 *
 * `CANONICAL_*` is read through {@link brandCanonical}, a seam a test can
 * override. That lets a harness test inject the post-rename value while
 * production still emits the pre-rename one — which is how a Wave-1 red can be
 * a *real* red rather than a self-certifying assertion.
 *
 * Residual-scan whitelist: this file and `src-tauri/src/brand.rs` are the only
 * two non-fixture files permitted to contain a legacy brand string.
 */

/** Every canonical brand identifier the app writes. */
export interface BrandCanonical {
  /** `ConversationRecordV2.createdBy` discriminant. */
  readonly createdBy: string
  /** Markdown fence language carrying an ACP plan payload. */
  readonly planFence: string
  /** Bundled dark color-theme id (also the default theme id). */
  readonly themeId: string
  /** Bundled light color-theme family id. */
  readonly themeFamilyLight: string
  /** `terminalUrlOpenMode` enum member selecting the built-in browser. */
  readonly urlOpenMode: string
  /** `localStorage` namespace prefix used by the web plugin-store stub. */
  readonly storagePrefix: string
  /** `localStorage` prefix for ad-hoc renderer keys (`<prefix>some-key`). */
  readonly storageKeyPrefix: string
  /** Prefix of every custom DOM event name the renderer dispatches. */
  readonly eventPrefix: string
  /** Prefix of every CSS custom property the renderer sets. */
  readonly cssVarPrefix: string
  /** Prefix of every global/element id injected into third-party pages. */
  readonly domGlobalPrefix: string
  /** Negotiated binary WebSocket subprotocol. */
  readonly wsSubprotocol: string
  /** Prefix of every environment variable the app reads. */
  readonly envPrefix: string
  /** Per-user-repository workspace directory name. */
  readonly workspaceDir: string
  /** Short display name; also the `~/Documents/<name>` path component. */
  readonly displayName: string
  /** Full display name / product name. */
  readonly displayNameFull: string
  /** Desktop binary and npm/cargo package name. */
  readonly packageName: string
  /** Standalone headless server binary name. */
  readonly serverBinary: string
  /** Production bundle identifier. */
  readonly bundleId: string
  /** Development bundle identifier. */
  readonly bundleIdDev: string
  /** Deep-link URL scheme (without `://`). */
  readonly deepLinkScheme: string
  /** Base file name (no extension) of the desktop log file. */
  readonly logFileName: string
  /** `log`/`tracing` target prefix and `RUST_LOG` directive stem. */
  readonly logTarget: string
  /** Keychain service holding desktop general credentials. */
  readonly keychainService: string
  /** Keychain service holding SSH passwords and key passphrases. */
  readonly keychainSshService: string
  /** Keychain service holding the iOS pairing secret. */
  readonly keychainPairingService: string
  /** MCP server name exposed to agents. */
  readonly mcpServerName: string
  /** Name of the managed scheduled-tasks agent skill. */
  readonly skillName: string
  /** HTML marker identifying a skill file this app wrote. */
  readonly skillMarker: string
  /**
   * On-disk key of the managed-skill manifest's ownership flag
   * (`.se-manager/managed-skills.json`).
   *
   * Mirrors `brand.rs`'s `skill_manifest_key`. The Rust struct that owns this
   * file carries `#[serde(rename_all = "camelCase", deny_unknown_fields)]`, so
   * the field identifier *is* the JSON key and there is no literal to grep for;
   * `deny_unknown_fields` means an existing user manifest stops deserializing
   * the moment the key moves. Held here so both sides keep an identical field
   * set — nothing in the renderer reads this file.
   */
  readonly skillManifestKey: string
  /** frp `[[proxies]]` registration name. */
  readonly frpProxyName: string
  /** Standalone-server state root directory name (unix, lowercase). */
  readonly stateDir: string
  /** iOS `UserDefaults` key prefix. */
  readonly iosDefaultsPrefix: string
  /** iOS on-disk cache directory name. */
  readonly iosCacheDir: string
}

/**
 * Values already written to user disks. Permanent, read-only, never re-emitted.
 *
 * Migration and compatibility-read paths are the *only* legitimate consumers.
 */
export const LEGACY: BrandCanonical = {
  createdBy: 'termul',
  planFence: 'termul-plan',
  themeId: 'termul',
  themeFamilyLight: 'termul-light',
  urlOpenMode: 'termul',
  storagePrefix: 'termul-store:',
  storageKeyPrefix: 'termul:',
  eventPrefix: 'termul:',
  cssVarPrefix: '--termul-',
  domGlobalPrefix: '__termul',
  wsSubprotocol: 'termul-terminal-v2.binary',
  envPrefix: 'TERMUL_',
  workspaceDir: '.termul',
  displayName: 'Termul',
  displayNameFull: 'Termul Manager',
  packageName: 'termul-manager',
  serverBinary: 'termul-server',
  bundleId: 'com.termul-manager.app',
  bundleIdDev: 'com.termul-manager.app.dev',
  deepLinkScheme: 'termul',
  logFileName: 'termul',
  logTarget: 'termul',
  keychainService: 'com.termul.manager',
  keychainSshService: 'termul-ssh',
  keychainPairingService: 'com.termul.remote.pairing',
  mcpServerName: 'termul',
  skillName: 'termul-scheduled-tasks',
  skillMarker: '<!-- managed-by-termul:termul-scheduled-tasks -->',
  skillManifestKey: 'managedByTermul',
  frpProxyName: 'termul',
  stateDir: 'termul',
  iosDefaultsPrefix: 'termul.',
  iosCacheDir: 'TermulRemote'
} as const

/**
 * Values the app writes today.
 *
 * Wave 5 flips these one contract at a time. Until a contract's flip task
 * lands, its entry here still equals the corresponding {@link LEGACY} value —
 * that gap is precisely what makes the Wave-1 harness tests go red.
 */
const DEFAULT_CANONICAL: BrandCanonical = {
  createdBy: 'se-manager',
  planFence: 'se-plan',
  themeId: 'se',
  themeFamilyLight: 'se-light',
  urlOpenMode: 'se',
  storagePrefix: 'se-store:',
  storageKeyPrefix: 'se:',
  eventPrefix: 'se:',
  cssVarPrefix: '--se-',
  domGlobalPrefix: '__se',
  wsSubprotocol: 'se-terminal-v2.binary',
  envPrefix: 'SE_',
  workspaceDir: '.se-manager',
  displayName: 'Se',
  displayNameFull: 'Se Manager',
  packageName: 'se-manager',
  serverBinary: 'se-server',
  bundleId: 'com.se-manager.app',
  bundleIdDev: 'com.se-manager.app.dev',
  deepLinkScheme: 'termul',
  logFileName: 'se-manager',
  logTarget: 'termul',
  keychainService: 'com.se-manager.app',
  keychainSshService: 'com.se-manager.ssh',
  keychainPairingService: 'com.termul.remote.pairing',
  mcpServerName: 'se-manager',
  skillName: 'se-manager-scheduled-tasks',
  skillMarker: '<!-- managed-by-se-manager:se-manager-scheduled-tasks -->',
  skillManifestKey: 'managedBySeManager',
  frpProxyName: 'se-manager',
  stateDir: 'se-manager',
  iosDefaultsPrefix: 'termul.',
  iosCacheDir: 'TermulRemote'
} as const

let override: Partial<BrandCanonical> | null = null

/**
 * The canonical brand values in force right now.
 *
 * Always call this rather than capturing a field at module scope — a value
 * captured into a top-level `const` freezes before a test can override it.
 */
export function brandCanonical(): BrandCanonical {
  return override ? { ...DEFAULT_CANONICAL, ...override } : DEFAULT_CANONICAL
}

/**
 * Test seam: force some canonical values for the duration of a test.
 *
 * Production never calls this. Harness tests use it to inject the *post*-rename
 * value while production still emits the pre-rename one, so the resulting red
 * reflects a real missing capability instead of a stale literal.
 */
export function __setBrandCanonicalOverride(next: Partial<BrandCanonical> | null): void {
  override = next
}

/** Restore the shipped canonical values. */
export function __resetBrandCanonicalOverride(): void {
  override = null
}

/**
 * The values a compatibility read must accept for `field`, most-current first.
 *
 * A read that knows only the canonical value drops everything already on disk;
 * one that knows only the legacy value stops working the moment the contract
 * flips. This is the one shape that survives both, and it collapses to a single
 * entry while the two spellings are still equal — so no caller pays for a
 * second lookup before the flip.
 *
 * Compatibility *read* only. A caller that writes must name the value it writes
 * explicitly; FORBID-04 forbids emitting two spellings of the same contract.
 *
 * Call it at the point of use. A list captured into a module-level `const`
 * freezes before {@link __setBrandCanonicalOverride} can move it.
 */
export function acceptedBrandValues<K extends keyof BrandCanonical>(field: K): readonly string[] {
  const canonical = brandCanonical()[field]
  return canonical === LEGACY[field] ? [canonical] : [canonical, LEGACY[field]]
}

/**
 * {@link acceptedBrandValues} as a regexp alternation body, ready to drop into
 * a `(?:…)` group.
 *
 * Every value is escaped: brand identifiers carry `.` (`com.termul.manager`)
 * and `-` prefixes, and an unescaped `.` in an alternation quietly matches any
 * character. Build the `RegExp` in the function body that uses it — a pattern
 * hoisted to module scope freezes with the brand value baked in.
 */
export function acceptedBrandPattern<K extends keyof BrandCanonical>(field: K): string {
  return acceptedBrandValues(field)
    .map((value) => value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'))
    .join('|')
}
